core/jvm/src/main/scala/zio/internal/ZScheduler.scala

```diff
@@ -511,6 +481,29 @@ private[zio] object ZScheduler {
     def put(trace: Trace): Long
   }
 
+  private object Locations {
+
+    final class Enabled(initialSize: Int = 64) extends Locations {
+      private[this] val locations = mutable.HashMap.empty[Trace, AtomicLong]
```

Comment:
The previous code was using a Java `HashMap`. I think It'd be better to continue with the Java one as it probably avoids some allocations (the thunk `=> default` in the `getOrElse` for example)

```scala
final class Enabled(initialSize: Int = 64) extends Locations {
      private[this] val locations = new java.util.HashMap[Trace, AtomicLong]((initialSize / 0.75d).toInt)

      def get(trace: Trace): Long = {
        val v = locations.get(trace)
        if (v eq null) 0L else v.get()
      }

      def put(trace: Trace): Long = {
        val v = locations.get(trace)
        if (v eq null) {
          locations.put(trace, new AtomicLong(1L))
          0L
        } else v.getAndIncrement()
      }
    }
```

Reply:
The problem with `java.util.HashMap` is that it's not thread-safe.

I think we can use a `java.util.concurrent.ConcurrentHashMap` though and utilize its `computeIfAbsent` method

Reply:
I did some more performance investigation on this. It seems that Scala's `mutable.HashMap` might be the better choice to use here:

1. It has a better read throughput than j.u.c.ConcurrentHashMap
2. The thunk doesn't cause an allocation when it doesn't depend on a value defined outside of

As you can see in the benchmarking code below, both implementations have roughly the same allocations, with Scala's mutable.HashMap being faster

Benchmark code

```scala
package example.benchmarks

import io.netty.util.internal.ThreadLocalRandom
import org.openjdk.jmh.annotations._

import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong
import scala.collection.mutable

@State(Scope.Benchmark)
@BenchmarkMode(Array(Mode.AverageTime))
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
@Threads(8)
class MapGetBenchmark {
  @Param(Array("10", "100", "1000"))
  var nKeys: Int = _

  var sm: mutable.HashMap[String, AtomicLong]                         = _
  var jcm: java.util.concurrent.ConcurrentHashMap[String, AtomicLong] = _

  @Setup(Level.Trial)
  def setup(): Unit = {
    sm = mutable.HashMap.empty[String, AtomicLong]
    sm.sizeHint(nKeys)
    jcm = new java.util.concurrent.ConcurrentHashMap[String, AtomicLong](((nKeys + 1) / 0.75d).toInt)
  }

  @Benchmark
  def scalaMap() = {
    val i = ThreadLocalRandom.current().nextInt(nKeys).toString
    sm.getOrElseUpdate(i, new AtomicLong())
  }

  private val func: java.util.function.Function[String, AtomicLong] = _ => new AtomicLong()
  @Benchmark
  def javaMap() = {
    val i = ThreadLocalRandom.current().nextInt(nKeys).toString
    jcm.computeIfAbsent(i, func)
  }
}
```

Results:

```
[info] Benchmark                                    (nKeys)  Mode  Cnt      Score      Error   Units
[info] MapGetBenchmark.javaMap                           10  avgt    5     12.946 ±    0.953   ns/op
[info] MapGetBenchmark.javaMap:gc.alloc.rate             10  avgt    5  28286.805 ± 2031.354  MB/sec
[info] MapGetBenchmark.javaMap:gc.alloc.rate.norm        10  avgt    5     48.000 ±    0.001    B/op
[info] MapGetBenchmark.javaMap:gc.count                  10  avgt    5    150.000             counts
[info] MapGetBenchmark.javaMap:gc.time                   10  avgt    5     90.000                 ms
[info] MapGetBenchmark.javaMap                          100  avgt    5     29.946 ±    0.172   ns/op
[info] MapGetBenchmark.javaMap:gc.alloc.rate            100  avgt    5  12219.120 ±   64.465  MB/sec
[info] MapGetBenchmark.javaMap:gc.alloc.rate.norm       100  avgt    5     48.000 ±    0.001    B/op
[info] MapGetBenchmark.javaMap:gc.count                 100  avgt    5    126.000             counts
[info] MapGetBenchmark.javaMap:gc.time                  100  avgt    5     61.000                 ms
[info] MapGetBenchmark.javaMap                         1000  avgt    5     50.835 ±    0.219   ns/op
[info] MapGetBenchmark.javaMap:gc.alloc.rate           1000  avgt    5   7198.896 ±   47.398  MB/sec
[info] MapGetBenchmark.javaMap:gc.alloc.rate.norm      1000  avgt    5     48.000 ±    0.001    B/op
[info] MapGetBenchmark.javaMap:gc.count                1000  avgt    5    100.000             counts
[info] MapGetBenchmark.javaMap:gc.time                 1000  avgt    5     46.000                 ms
[info] MapGetBenchmark.scalaMap                          10  avgt    5      9.361 ±    0.140   ns/op
[info] MapGetBenchmark.scalaMap:gc.alloc.rate            10  avgt    5  19543.957 ±  319.883  MB/sec
[info] MapGetBenchmark.scalaMap:gc.alloc.rate.norm       10  avgt    5     24.000 ±    0.001    B/op
[info] MapGetBenchmark.scalaMap:gc.count                 10  avgt    5    132.000             counts
[info] MapGetBenchmark.scalaMap:gc.time                  10  avgt    5    100.000                 ms
[info] MapGetBenchmark.scalaMap                         100  avgt    5     17.891 ±    2.119   ns/op
[info] MapGetBenchmark.scalaMap:gc.alloc.rate           100  avgt    5  10658.125 ±  715.312  MB/sec
[info] MapGetBenchmark.scalaMap:gc.alloc.rate.norm      100  avgt    5     24.000 ±    0.001    B/op
[info] MapGetBenchmark.scalaMap:gc.count                100  avgt    5    118.000             counts
[info] MapGetBenchmark.scalaMap:gc.time                 100  avgt    5     57.000                 ms
[info] MapGetBenchmark.scalaMap                        1000  avgt    5     28.983 ±    0.064   ns/op
[info] MapGetBenchmark.scalaMap:gc.alloc.rate          1000  avgt    5  12628.195 ±   31.627  MB/sec
[info] MapGetBenchmark.scalaMap:gc.alloc.rate.norm     1000  avgt    5     48.000 ±    0.001    B/op
[info] MapGetBenchmark.scalaMap:gc.count               1000  avgt    5    140.000             counts
[info] MapGetBenchmark.scalaMap:gc.time                1000  avgt    5     66.000                 ms
```